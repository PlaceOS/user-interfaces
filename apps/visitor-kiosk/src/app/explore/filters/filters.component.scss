@import 'variables';

:host {
    display: flex;
    align-items: center;
    flex-direction: column;
    padding: 1em;
    min-width: 24em;
}

mat-checkbox {
    width: 100%;
    font-size: 1.25em;
}

button {
    min-width: 8em;
    margin-top: 1em;
}

header {
    font-size: 1.5em;
    margin-bottom: .5em;
    padding: .5em;
    text-align: center;
    width: 100%;
    color: #fff;
    background-color: $color-primary;
    border-radius: 4px;
    font-family: $heading-font;
}
