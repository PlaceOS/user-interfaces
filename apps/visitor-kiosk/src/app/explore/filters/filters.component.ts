import { Component } from '@angular/core';
import { MatDialogRef } from '@angular/material/dialog';

import { Identity } from '@user-interfaces/common';

@Component({
    selector: 'filter-list',
    templateUrl: './filters.component.html',
    styleUrls: ['./filters.component.scss']
})
export class FiltersComponent {
    /** Type of item being filtered */
    public name: string;
    /** List of available filter options */
    public filters: Identity[] = [];

    constructor(private _dialog_ref: MatDialogRef<FiltersComponent>) { }

    public applyChanges() {
        this._dialog_ref.close();
    }
}
