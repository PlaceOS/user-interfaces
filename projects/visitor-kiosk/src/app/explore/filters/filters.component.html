<header>{{ name }} Filters</header>
<mat-checkbox *ngFor="let item of filters" [checked]="item.active" (change)="item.active = $event.checked">
    {{ item.name }}
</mat-checkbox>
<button mat-button (click)="applyChanges()">Apply</button>
