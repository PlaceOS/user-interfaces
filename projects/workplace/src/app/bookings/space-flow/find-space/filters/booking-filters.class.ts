import { Building } from 'src/app/organisation/building.class';
import { unique } from 'src/app/common/general';
import { RoomConfiguration } from 'src/app/common/room-configuration.interface';
import { Identity } from 'src/app/common/types';

export type BookingFilterKey = 'locations' | 'configurations' | 'features' | 'capacity';
export type BookingFilterValue = Building | RoomConfiguration | Identity | null;

export class BookingFilters {
    constructor(
        public locations: string[] = [],
        public configurations: string[] = [],
        public features: string[] = [],
        public capacity: string = null
    ) {}

    public setItem(key: BookingFilterKey, val: BookingFilterValue): BookingFilters {
        const newFilter = this.clone();
        if (!!val) {
            switch (key) {
                case 'capacity':
                    newFilter.capacity = (val as Identity).value;
                    break;
                default: {
                    // a nasty cast but RoomConfiguration, Building and Identity all have an id field
                    const { id } = val as any;
                    let array = newFilter[key];
                    array.push(id);
                    array = unique(array);
                    newFilter[key] = [...array];
                    break;
                }
            }
        }
        return newFilter;
    }

    public removeItem(key: BookingFilterKey, val: BookingFilterValue): BookingFilters {
        const newFilter = this.clone();
        if (!!val) {
            switch (key) {
                case 'capacity':
                    newFilter.capacity = null;
                    break;
                default: {
                    // a nasty cast but RoomConfiguration, Building and Identity all have an id field
                    const { id } = val as any;
                    const array = newFilter[key];
                    array.splice(
                        array.findIndex(objId => objId === id),
                        1
                    );
                    newFilter[key] = [...array];
                    break;
                }
            }
        }
        return newFilter;
    }

    public get any(): boolean {
        return (
            !!this.locations.length ||
            !!this.configurations.length ||
            !!this.features.length ||
            !!this.capacity
        );
    }

    private clone(): BookingFilters {
        return new BookingFilters(
            this.locations,
            this.configurations,
            this.features,
            this.capacity
        );
    }
}
