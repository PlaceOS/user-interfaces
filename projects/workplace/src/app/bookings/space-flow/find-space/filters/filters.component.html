<div class="space-filters" *ngIf="date">
    <div class="actions">
        <a-date-field
            name="date"
            [ngModel]="date.value"
            (ngModelChange)="date.setValue($event)"
        ></a-date-field>
        <a-time-field
            name="start-time"
            [ngModel]="date.value"
            (ngModelChange)="date.setValue($event)"
        ></a-time-field>
        <an-action-field placeholder="Add Filter" [matMenuTriggerFor]="menu"></an-action-field>
        <mat-menu #menu>
            <button mat-menu-item [matMenuTriggerFor]="location_menu">
                <app-icon [icon]="{ class: 'material-icons', content: 'room' }"></app-icon>
                <div class="text">Location</div>
            </button>
            <button mat-menu-item [matMenuTriggerFor]="config_menu">
                <app-icon [icon]="{ class: 'material-icons', content: 'airline_seat_recline_normal' }"></app-icon>
                <div class="text">Room Config</div>
            </button>
            <button mat-menu-item [matMenuTriggerFor]="capacity_menu">
                <app-icon [icon]="{ class: 'material-icons', content: 'account_circle' }"></app-icon>
                <div class="text">Room Capacity</div>
            </button>
            <button mat-menu-item [matMenuTriggerFor]="features_menu">
                <app-icon [icon]="{ class: 'material-icons', content: 'devices_other' }"></app-icon>
                <div class="text">Room Features</div>
            </button>
        </mat-menu>
    </div>
    <div class="filters" *ngIf="hasFilters">
        <mat-chip-list aria-label="Location Filters">
            <mat-chip
                *ngFor="let building of location_list"
                [id]="building.id"
                color="primary"
                [removable]="true"
                (removed)="removeBookingFilter('locations', building)"
            >
                {{ building.name }}
                <app-icon
                    matChipRemove
                    [icon]="{ type: 'icon', class: 'material-icons', content: 'close' }"
                ></app-icon>
            </mat-chip>
            <mat-chip
                *ngFor="let obj of configurations_list"
                [id]="obj.id"
                color="primary"
                [removable]="true"
                (removed)="removeBookingFilter('configurations', obj)"
            >
                {{ obj.name }}
                <app-icon
                    matChipRemove
                    [icon]="{ type: 'icon', class: 'material-icons', content: 'close' }"
                ></app-icon>
            </mat-chip>
            <mat-chip
                *ngIf="!!selected_capacity"
                [id]="selected_capacity.value"
                color="primary"
                [removable]="true"
                (removed)="removeBookingFilter('capacity', selected_capacity)"
            >
                {{ selected_capacity.display }}
                <app-icon
                    matChipRemove
                    [icon]="{ type: 'icon', class: 'material-icons', content: 'close' }"
                ></app-icon>
            </mat-chip>
            <mat-chip
                *ngFor="let obj of features_list"
                [id]="obj.id"
                color="primary"
                [removable]="true"
                (removed)="removeBookingFilter('features', obj)"
            >
                {{ obj.name }}
                <app-icon
                    matChipRemove
                    [icon]="{ type: 'icon', class: 'material-icons', content: 'close' }"
                ></app-icon>
            </mat-chip>

        </mat-chip-list>
    </div>
</div>
<mat-menu #location_menu>
    <button mat-menu-item *ngFor="let item of building_list" (click)="setBookingFilter('locations', item)">
        {{ item.name }}
    </button>
</mat-menu>
<mat-menu #config_menu>
    <button mat-menu-item *ngFor="let item of building_room_configs" (click)="setBookingFilter('configurations', item)">
        {{ item.name }}
    </button>
</mat-menu>
<mat-menu #capacity_menu>
    <button mat-menu-item *ngFor="let item of capacity_filters" (click)="setBookingFilter('capacity', item)">
        {{ item.display }}
    </button>
</mat-menu>
<mat-menu #features_menu>
    <button mat-menu-item *ngFor="let item of building_room_features" (click)="setBookingFilter('features', item)">
        {{ item.name }}
    </button>
</mat-menu>
