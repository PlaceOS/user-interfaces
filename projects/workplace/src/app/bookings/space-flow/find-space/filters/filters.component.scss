
@import 'variables';
@import 'mixins';

.space-filters {
    border-bottom: 1px solid #ccc;
    width: 100%;
}

.actions {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 1em;

    @include respond-to(mobile) {
        padding: .5em;
        justify-content: flex-start;
    }

    > * {
        margin: 0 .5em;

        @include respond-to(mobile) {
            margin: 0 2px;
            min-width: 32%;
        }
    }
}

.filters {
    padding: 0 1em 1em;

    @include respond-to(mobile) {
        padding-bottom: .5em;
    }
}

button {
    margin-bottom: .2em !important;
}

.mat-menu-item {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0 1em;

    app-icon {
        font-size: 1.2em;
    }

    .text {
        flex: 1;
        min-width: 50%;
        margin-left: .5em;
        padding-right: 2em;
    }
}
