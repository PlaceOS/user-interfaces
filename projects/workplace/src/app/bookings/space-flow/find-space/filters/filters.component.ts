import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { FormControl } from '@angular/forms';
import { first } from 'rxjs/operators';

import { BaseClass } from 'src/app/common/base.class';
import { Building } from 'src/app/organisation/building.class';
import { RoomConfiguration } from 'src/app/common/room-configuration.interface';
import { Identity } from 'src/app/common/types';
import { BookingFilters, BookingFilterKey, BookingFilterValue } from './booking-filters.class';
import { OrganisationService } from 'src/app/organisation/organisation.service';
import { SettingsService } from 'src/app/common/settings.service';

@Component({
    selector: 'booking-space-filters',
    templateUrl: './filters.component.html',
    styleUrls: ['./filters.component.scss'],
})
export class BookingSpaceFiltersComponent extends BaseClass implements OnInit {
    /** Date selected for the booking */
    @Input() public date: FormControl;
    @Input() public bookingFilters: BookingFilters = new BookingFilters();

    /** Emitter for changes to filters */
    @Output() public bookingFiltersChange = new EventEmitter<BookingFilters>();

    /** List of available buildings */
    public building_list: Building[] = [];

    public building_room_configs: RoomConfiguration[] = [];

    public building_room_features: Identity[] = [];

    public capacity_filters: Identity[] = [];

    public get configurations_list(): RoomConfiguration[] {
        return (this.bookingFilters.configurations || []).map((id) =>
            this.building_room_configs.find((cfg) => cfg.id === id)
        );
    }

    public get features_list(): Identity[] {
        return (this.bookingFilters.features || []).map((id) =>
            this.building_room_features.find((feat) => feat.id === id)
        );
    }

    public get selected_capacity(): Identity | null {
        if (!this.bookingFilters.capacity) {
            return null;
        }
        return this.capacity_filters.find((f) => f.value === this.bookingFilters.capacity);
    }

    /** List of filtered location */
    public get location_list(): Building[] {
        return (this.bookingFilters.locations || []).map((id) =>
            this.building_list.find((bld) => bld.id === id)
        );
    }

    constructor(private _org: OrganisationService, private _settings: SettingsService) {
        super();
    }

    public ngOnInit(): void {
        this._org.initialised.pipe(first((_) => _)).subscribe(() => {
            this.building_list = this._org.buildings;
            this.building_room_configs = this.building_list
                .map((m) => m.room_configurations)
                .reduce((prev, curr) => prev.concat(curr))
                .filter((el, idx, arr) => arr.findIndex((f) => f.id === el.id) === idx);
            this.building_room_features = this.building_list
                .map((m) => m.extras)
                .reduce((prev, curr) => prev.concat(curr))
                .filter((el, idx, arr) => arr.findIndex((f) => f.id === el.id) === idx);
            this.capacity_filters = this._settings.get('app.booking.capacity_filters') || [];
            console.log(this.capacity_filters);
        });
    }

    public get hasFilters(): boolean {
        return this.bookingFilters.any;
    }

    public setBookingFilter(key: BookingFilterKey, val: BookingFilterValue) {
        this.bookingFilters = this.bookingFilters.setItem(key, val);
        this.bookingFiltersChange.emit(this.bookingFilters);
    }

    public removeBookingFilter(key: BookingFilterKey, val: BookingFilterValue) {
        this.bookingFilters = this.bookingFilters.removeItem(key, val);
        this.bookingFiltersChange.emit(this.bookingFilters);
    }
}
