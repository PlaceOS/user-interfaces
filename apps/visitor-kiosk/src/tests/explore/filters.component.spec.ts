import { MatCheckboxModule } from '@angular/material/checkbox';
import { MatDialogRef } from '@angular/material/dialog';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { FiltersComponent } from '../../app/explore/filters.component';

describe('FiltersComponent', () => {
    let spectator: Spectator<FiltersComponent>;
    const createComponent = createComponentFactory({
        component: FiltersComponent,
        providers: [{ provide: MatDialogRef, useValue: { close: jest.fn() } }],
        imports: [MatCheckboxModule],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
    });
});
