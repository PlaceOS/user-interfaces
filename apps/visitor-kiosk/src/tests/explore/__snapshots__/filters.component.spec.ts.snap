// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <header>
     Filters
  </header>
  
  <button
    mat-button=""
  >
    Apply
  </button>
</div>
`;
